exp/etcdrestore/webhooks/rke2config.go

```diff
@@ -45,6 +66,431 @@ func (r *RKE2ConfigWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a RKE2Config but got a %T", obj))
+	}
+
+	planSecretName := strings.Join([]string{rke2Config.Name, "rke2config", "plan"}, "-")
+
+	if err := r.createSecretPlanResources(ctx, planSecretName, rke2Config); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to create secret plan resources: %s", err))
+	}
+
+	serviceAccountToken, err := r.EnsureServiceAccountSecretPopulated(ctx, planSecretName)
+	if err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to ensure service account secret is populated: %s", err))
+	}
+
+	logger.Info("Service account secret is populated")
+
+	serverUrlSetting := &unstructured.Unstructured{}
+	serverUrlSetting.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "management.cattle.io",
+		Kind:    "Setting",
+		Version: "v3",
+	})
+
+	if err := r.Get(context.Background(), client.ObjectKey{
+		Name: "server-url",
+	}, serverUrlSetting); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to get server url setting: %s", err))
+	}
+	serverUrl, ok := serverUrlSetting.Object["value"].(string)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to get server url setting: %s", err))
+	}
+
+	if serverUrl == "" {
+		return apierrors.NewBadRequest("server url setting is empty")
+	}
+
+	caSetting := &unstructured.Unstructured{}
+	caSetting.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "management.cattle.io",
+		Kind:    "Setting",
+		Version: "v3",
+	})
+	if err := r.Get(context.Background(), client.ObjectKey{
+		Name: "cacerts",
+	}, caSetting); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to get ca setting: %s", err))
+	}
+
+	pem, ok := caSetting.Object["value"].(string)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to get ca setting: %s", err))
+	}
+
+	if err := r.CreateConnectInfoJson(ctx, rke2Config, planSecretName, serverUrl, pem, serviceAccountToken); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to create connect info json: %s", err))
+	}
+
+	if err := r.CreateSystemAgentInstallScript(ctx, serverUrl, rke2Config); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to create system agent install script: %s", err))
+	}
+
+	if err := r.CreateConfigYAML(rke2Config); err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("failed to create config.yaml: %s", err))
+	}
+
+	r.AddPostInstallCommands(rke2Config)
+
+	return nil
+}
+
+// createSecretPlanResources creates the secret, role, rolebinding, and service account for the plan.
+func (r *RKE2ConfigWebhook) createSecretPlanResources(ctx context.Context, planSecretName string, rke2Config *bootstrapv1.RKE2Config) error {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Creating secret plan resources")
+
+	var errs []error
+
+	// Create the ServiceAccount first to later pass to the RoleBinding creation
+	sa := r.createServiceAccount(planSecretName, rke2Config)
```

Comment:
We can create it later as we already know the namespace and name to use

Reply:
No we don't know it beforehand. I had to create it first and later refer to it.

Reply:
if `sa.Name` matches `planSecretName`  can we use `planSecretName` later instead?